File: q100-q199/q100-109/q108/q108.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        def sortedArrayToBST_helper(left: int, right: int) -> TreeNode:
            if left > right:
                return None
            middle = (left + right) // 2
            node = TreeNode(nums[middle])
            node.left = sortedArrayToBST_helper(left, middle - 1)
            node.right = sortedArrayToBST_helper(middle + 1, right)
            return node
        return sortedArrayToBST_helper(0, len(nums) - 1)

File: Sequence/q100-q199/q100-109/q108/q108.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return sortedArrayToBST_helper(nums, 0, nums.size() - 1);
    }
    
    TreeNode* sortedArrayToBST_helper(vector<int>& nums, int left, int right) {
        if (left > right) {
            return NULL;
        }
        int middle = (left + right) / 2;
        TreeNode* node = new TreeNode(nums[middle]);
        node->left = sortedArrayToBST_helper(nums, left, middle - 1);
        node->right = sortedArrayToBST_helper(nums, middle + 1, right);
        return node;
    }
};
